tests/unit-tests/installer/test-class-installer.php

```diff
@@ -0,0 +1,336 @@
+
+		/* Assert. */
+		$actual_flag = get_option( 'sensei_enrolment_legacy' );
+		$this->assertNotEmpty( $actual_flag, 'Legacy update flag option should be set during v1 updates without progress artifacts' );
+	}
+
+	public function testUpdate_WhenCurrentVersionIsV2AndWithoutProgress_SetsExpectedVersion(): void {
+		/* Arrange. */
+		$this->resetUpdateOptions();
+
+		update_option( 'sensei-version', '2.4.0' );
+
+		/* Act. */
+		Installer::instance( SENSEI_LMS_VERSION )->update();
+
+		/* Assert. */
+		$actual_version = get_option( 'sensei-version' );
+		$this->assertEquals( Sensei()->version, $actual_version );
+	}
+
+	public function testUpdate_WhenCurrentVersionIsV2AndWithoutProgress_SetsUpdateLegacyFlag(): void {
+		/* Arrange. */
+		$this->resetUpdateOptions();
+
+		update_option( 'sensei-version', '2.4.0' );
+
+		/* Act. */
+		Installer::instance( SENSEI_LMS_VERSION )->update();
+
+		/* Assert. */
+		$actual_flag = get_option( 'sensei_enrolment_legacy' );
+		$this->assertNotEmpty( $actual_flag, 'Legacy update flag option should be set during v2 updates with known previous version' );
+	}
+
+	public function testUpdate_WhenCurrentVersionIsV2AndWithProgress_SetsExpectedVersion(): void {
+		/* Arrange. */
+		$user_id   = $this->factory->user->create();
+		$course_id = $this->factory->course->create();
+		Sensei_Utils::user_start_course( $user_id, $course_id );
+
+		$this->resetUpdateOptions();
+
+		update_option( 'sensei-version', '2.4.0' );
+
+		/* Act. */
+		Installer::instance( SENSEI_LMS_VERSION )->update();
+
+		/* Assert. */
+		$actual_version = get_option( 'sensei-version' );
+		$this->assertEquals( Sensei()->version, $actual_version );
+	}
+
+	public function testUpdate_WhenCurrentVersionIsV2AndWithProgress_SetsLegacyUpdateFlag(): void {
+		/* Arrange. */
+		$user_id   = $this->factory->user->create();
+		$course_id = $this->factory->course->create();
+		Sensei_Utils::user_start_course( $user_id, $course_id );
+
+		$this->resetUpdateOptions();
+
+		update_option( 'sensei-version', '2.4.0' );
+
+		/* Act. */
+		Installer::instance( SENSEI_LMS_VERSION )->update();
+
+		/* Assert. */
+		$actual_flag = get_option( 'sensei_enrolment_legacy' );
+		$this->assertNotEmpty( $actual_flag, 'Legacy update flag option should be set during v2 updates with progress' );
+	}
+
+	public function testUpdate_WhenInitHookWasTriggered_CallsTheMethod(): void {
+		/* Arrange. */
+		global $wp_filter;
+		$initial_wp_filter = $wp_filter;
+		$wp_filter         = [];
+
+		$updates         = $this->createMock( \Sensei_Updates::class );
+		$updates_factory = $this->createMock( Updates_Factory::class );
+		$updates_factory->method( 'create' )->willReturn( $updates );
+
+		$installer = new Installer( new Schema(), $updates_factory, SENSEI_LMS_VERSION );
+		$installer->init();
+
+		/* Expect & Act. */
+		$updates->expects( $this->once() )->method( 'run_updates' );
+		do_action( 'init' );
+
+		$wp_filter = $initial_wp_filter;
+	}
```

Comment:
I think this could be simplified. 🙂 What do you think of the following:

```suggestion
	public function testInit_WhenCalled_RegistersHooks(): void {
		/* Arrange. */
		$installer = new Installer( new Schema(), new Updates_Factory(), SENSEI_LMS_VERSION );
		$installer->init();

		/* Assert. */
		$this->assertSame( 10, has_action( 'plugins_loaded', [ $installer, 'install' ] ) );
		$this->assertSame( 10, has_action( 'init', [ $installer, 'update' ] ) );
	}
```

Do you think this would be enough?

Reply:
Or maybe that suggestion could be added as a separate test just to make sure the hooks are there?

Reply:
Yup, I thought about keeping both :) As they test different things :)

Reply:
Added a test here: https://github.com/Automattic/sensei/pull/5857/commits/df230e043a9c7acb440e3767e30482073dd9bd95